File: src/router/bmon/trem/graph.php
<?
/*
 * Copyright (c) 2001-2005 Thomas Graf <tgraf@suug.ch>
 *
 * Permission is hereby granted, free of charge, to any person obtaining a
 * copy of this software and associated documentation files (the "Software"),
 * to deal in the Software without restriction, including without limitation
 * the rights to use, copy, modify, merge, publish, distribute, sublicense,
 * and/or sell copies of the Software, and to permit persons to whom the
 * Software is furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included
 * in all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
 * OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
 * THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
 * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
 * DEALINGS IN THE SOFTWARE.
 */

include "config.php";
include "charts.php";

$table = "";
$attr = $_GET["attr"];
$unit = $_GET["unit"];
$limit = $_GET["limit"];
$type = $_GET["type"];
$update = $_GET["update"];
$from_method = $_GET["from_method"];
$from_ts = $_GET["from_ts"];
$to_method = $_GET["to_method"];
$to_ts = $_GET["to_ts"];
$gwidth = $_GET["gwidth"];
$gheight = $_GET["gheight"];

if (!isset($attr) || !is_numeric($attr))
	return;

if (isset($limit)) {
	if (!is_numeric($limit))
		return;
} else
	$limit = 60;

if (!isset($gwidth) || $gwidth < 200)
	$gwidth = 550;

if (!isset($gheight) || $gheight < 50)
	$gheight = 150;

if (!isset($unit))
	return;

if (!isset($type) ||
    ($type != "l" && $type != "c" && $type != "sc" && $type != "a" &&
     $type != "sa"))
	$type = "l";

if (isset($update) && !is_numeric($update))
	return;

if (!isset($update) || $update < 0 || $update > 31536000)
	$update = 0;

if (isset($from_method) &&
    ($from_method != "auto" && $from_method != "sel"))
	return;

if (!isset($from_method))
	$from_method = "auto";

if (isset($to_method) &&
    ($to_method != "auto" && $to_method != "sel"))
	return;

if (!isset($to_method))
	$to_method = "auto";

if ($unit == "s")
	$table = "hist_s";
else if ($unit == "m")
	$table = "hist_m";
else if ($unit == "h")
	$table = "hist_h";
else if ($unit == "d")
	$table = "hist_d";
else
	return;

if ($type == "l")
	$rtype = "line";
else if ($type == "c")
	$rtype = "column";
else if ($type == "sc")
	$rtype = "stacked column";
else if ($type == "a")
	$rtype = "area";
else if ($type == "sa")
	$rtype = "stacked area";
else
	return;

mysql_connect($db_host, $db_user, $db_password);
mysql_select_db($db_name);

$query = "SELECT (concat(ts, '.', offset) - 0.0) ts, rx_rate, tx_rate
	FROM $table WHERE attr = " . mysql_escape_string($attr);

if ($from_method == "sel" && isset($from_ts))
	$query .= " AND ts > ". mysql_escape_string($from_ts);

if ($to_method == "sel" && isset($to_ts))
	$query .= " AND ts < ". mysql_escape_string($to_ts);

$query .= " ORDER BY ts DESC LIMIT " . mysql_escape_string($limit);

$r = mysql_query($query);
if (!$r) {
	print "<b>Database Error: " . mysql_error() . "</b><br>\n";
	return 1;
}

$a_ts = array();
$a_rx = array();
$a_tx = array();

$i = 0;
$max = 0;
$nr_cols = 0;

while (list($ts, $rx, $tx) = mysql_fetch_row($r)) {
	$nr_cols++;
	if ($unit == "s")
		array_push($a_ts, strftime("%H:%M:%S", $ts));
	else if ($unit == "m")
		array_push($a_ts, strftime("%H:%M:%S", $ts));
	else if ($unit == "h")
		array_push($a_ts, strftime("%H:%M", $ts));
	else if ($unit == "d")
		array_push($a_ts, strftime("%m/%d", $ts));

	if ($rx == 4294967295)
		array_push($a_rx, null);
	else
		array_push($a_rx, $rx);

	if ($tx == 4294967295)
		array_push($a_tx, null);
	else
		array_push($a_tx, $tx);

	if ($rx && $rx > $max)
		$max = $rx;
	if ($tx && $tx > $max)
		$max = $tx;
}

$divisor = find_divisor($max * 0.8);
$u = find_unit($divisor);

for ($i = 0; $i < count($a_rx); $i++)
	if ($a_rx[$i])
		$a_rx[$i] = $a_rx[$i] / $divisor;

for ($i = 0; $i < count($a_tx); $i++)
	if ($a_tx[$i])
		$a_tx[$i] = $a_tx[$i] / $divisor;

array_push($a_ts, "");
array_push($a_rx, "RX");
array_push($a_tx, "TX");

$chart["axis_category"] = array('skip'	=> (($nr_cols / 6) - 1),
				'bold'	=> false,
				'size'	=> 10);

$chart["axis_ticks"] = array('value_ticks'	=> true,
			     'category_ticks'	=> true,
			     'minor_count'	=> 4);

$chart["axis_value"] = array('decimals'		=> 0,
			     'bold'		=> false,
			     'size'		=> 10,
			     'suffix'		=> " $u",
			     'show_min'		=> false);

$chart["chart_bg"] = array('positive_color'	=> "EEEEEE");

$chart["chart_pref"] = array('line_thickness'	=> 1,
			     'point_shape'	=> "none");

if ($update > 0) {
	$chart["live_update"] = array('url'   => "$PHP_SELF?attr=$attr&unit=$unit&limit=$limit&type=$type&update=$update&time=".time(),
				      'delay' => $update);
}

$chart["chart_border"] = array('bottom_thickness' => 0);

$chart["chart_data"] = array(array_reverse($a_ts),
	array_reverse($a_rx), array_reverse($a_tx));

$chart["chart_rect"] = array('width' => $gwidth - 80,
			     'x' => $gheight - 90);

$chart["chart_value"] = array('position' => "cursor",
			      'bold' => false,
			      'size' => 12);

$chart["legend_rect"] = array('width' => 120,
			      'x' => ($gwidth / 2) - 40,
			      'y' => 5,
			      'height' => 10);

$chart["legend_label"] = array('bullet' => "line",
			       'size' => 11);

$chart["legend_bg"] = array('bg_alpha' => 0);

$chart["series_gap"] = array('bar_gap'	=> 0,
			     'set_gap'  => 0);

$chart["chart_type"] = $rtype;

mysql_close();

SendChartData($chart);
?>

File: src/router/bmon/src/graph.c
/*
 * graph.c             Graph creation utility
 *
 * Copyright (c) 2001-2005 Thomas Graf <tgraf@suug.ch>
 *
 * Permission is hereby granted, free of charge, to any person obtaining a
 * copy of this software and associated documentation files (the "Software"),
 * to deal in the Software without restriction, including without limitation
 * the rights to use, copy, modify, merge, publish, distribute, sublicense,
 * and/or sell copies of the Software, and to permit persons to whom the
 * Software is furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included
 * in all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
 * OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
 * THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
 * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
 * DEALINGS IN THE SOFTWARE.
 */

#include <bmon/bmon.h>
#include <bmon/graph.h>
#include <bmon/input.h>
#include <bmon/conf.h>
#include <bmon/item.h>
#include <bmon/node.h>
#include <bmon/conf.h>
#include <bmon/utils.h>


static void put_col(table_t *t, int data_idx, hist_data_t *src, int hist_idx,
		    double half_step)
{
	int i;
	char *col = D_AT_COL(t->t_data, data_idx);

	rate_cnt_t tot = src->hd_data[hist_idx];

	if (tot == UNK_DATA) {
		for (i = 0; i < t->t_height; i++)
			*(D_AT_ROW(col, i)) = get_unk_char();
	} else if (tot) {
		*(D_AT_ROW(col, 0)) = ':';
		
		for (i = 0; i < t->t_height; i++)
			if (tot >= (t->t_y_scale[i] - half_step))
				*(D_AT_ROW(col, i)) = get_fg_char();
	}
}

static void create_table(table_t *t, hist_data_t *src, int index, int height,
			 int unit)
{
	int i, di;
	size_t dsize = height * (HISTORY_SIZE + 1);
	rate_cnt_t max = 0;
	double half_step, step;
	
	t->t_index = index;
	t->t_height = height;
	t->t_y_scale = xcalloc(height, sizeof(double));
	t->t_data = xcalloc(dsize, sizeof(char));

	memset(t->t_data, get_bg_char(), dsize);

	for (i = 0; i < height; i++)
		*(D_AT_COL(D_AT_ROW(t->t_data, i), HISTORY_SIZE)) = '\0';

	for (i = 0; i < HISTORY_SIZE; i++)
		if (max < src->hd_data[i] && src->hd_data[i] != UNK_DATA)
			max = src->hd_data[i];

	step = (double) max / (double) height;
	half_step = step / 2.0f;

	for (i = 0; i < height; i++)
		t->t_y_scale[i] = (double) (i + 1) * step;

	for (di = 0, i = (index - 1); i >= 0; di++, i--)
		put_col(t, di, src, i, half_step);

	for (i = (HISTORY_SIZE - 1); di < HISTORY_SIZE; di++, i--)
		put_col(t, di, src, i, half_step);

	{
		b_cnt_t div;
		int h = (height / 3) * 2;

		if (h >= height)
			h = (height - 1);

		div = get_divisor(t->t_y_scale[h], unit, &t->t_y_unit, NULL);
		
		for (i = 0; i < height; i++)
			t->t_y_scale[i] /= (double) div;
	}
}

graph_t * create_graph(hist_elem_t *src, int height, int unit)
{
	graph_t *g;

	g = xcalloc(1, sizeof(graph_t));

	create_table(&g->g_rx, &src->he_rx, src->he_index, height, unit);
	create_table(&g->g_tx, &src->he_tx, src->he_index, height, unit);

	return g;
}

graph_t * create_configued_graph(history_t *src, int height, int unit,
				 int x_unit)
{
	graph_t *g;
	hist_elem_t *e = NULL;
	char *u = "s";
	int h = 0;

	switch (x_unit) {
		case X_SEC:  u = "s"; e = &src->h_sec; break;
		case X_MIN:  u = "m"; e = &src->h_min; break;
		case X_HOUR: u = "h"; e = &src->h_hour; break;
		case X_DAY:  u = "d"; e = &src->h_day; break;
		case X_READ: {
			if (get_read_interval() != 1.0f) {
				char buf[32];
				float ri = get_read_interval();

				snprintf(buf, sizeof(buf), "(%.2fs)", ri);
				u = strdup(buf);
				h = 1;
				e = &src->h_read;
			} else {
				u = "s";
				e = &src->h_sec;
			}
		}
		break;
	}

	if (NULL == e)
		BUG();

	g = create_graph(e, height, unit);

	if (h)
		g->g_flags |= GRAPH_HAS_FREEABLE_X_UNIT;

	g->g_rx.t_x_unit = u;
	g->g_tx.t_x_unit = u;

	return g;
}

static void free_table(table_t *t)
{
	xfree(t->t_y_scale);
	xfree(t->t_data);
}

void free_graph(graph_t *g)
{
	if (g->g_flags & GRAPH_HAS_FREEABLE_X_UNIT)
		xfree(g->g_rx.t_x_unit);
		
	free_table(&g->g_rx);
	free_table(&g->g_tx);
	xfree(g);
}

void new_graph(void)
{
	if (get_ngraphs() >= (MAX_GRAPHS - 1))
		return;
	set_ngraphs(get_ngraphs() + 1);
}

void del_graph(void)
{
	if (get_ngraphs() <= 1)
		return;
	set_ngraphs(get_ngraphs() - 1);
}

int next_graph(void)
{
	item_t *it = get_current_item();
	if (it == NULL)
		return EMPTY_LIST;

	if (it->i_graph_sel >= (get_ngraphs() - 1))
		it->i_graph_sel = 0;
	else
		it->i_graph_sel++;

	return 0;
}
	
int prev_graph(void)
{
	item_t *it = get_current_item();
	if (it == NULL)
		return EMPTY_LIST;

	if (it->i_graph_sel <= 0)
		it->i_graph_sel = get_ngraphs() - 1;
	else
		it->i_graph_sel--;

	return 0;
}
